soc: arm: atmel: implement SAM V70 support

iperry opened the pull request:
This patchset implements support for the Atmel SAM V70(b) series devices.

Just the basics, soc/ and DTS bindings.

Requires https://github.com/zephyrproject-rtos/hal_atmel/pull/23

Signed-off-by: Perry Hung <perry@genrad.io>

iperry committed:
dts: arm: atmel: add samv70 dts bindings

Add DTS bindings for the SAM V70(b) device family.

Signed-off-by: Perry Hung <perry@genrad.io>

iperry committed:
soc: samv70: add SAM V70 soc support

Implement soc files for the samv70(b) family devices.

Signed-off-by: Perry Hung <perry@genrad.io>

iperry committed:
west.yml: hal_atmel: import ASF headers for SAMV70

Import upstream ASF support pack for the SAM V70(b) series devices.

Signed-off-by: Perry Hung <perry@genrad.io>

nandojve commented:
Hi @iperry ,

To move with this we need a board to build at CI at least. As far I remember, there is no official board available from MCHP that uses SAMV70. Once SAMV70 it is a subset from SAMV71 (less Ethernet) I`ll suggest to add entries at boards/arm/sam_v71_xult for that purpose. In this case, there will be have 4 targets inside sam_v71_xult board directory.

iperry commented:
> Hi @iperry ,
> 
> To move with this we need a board to build at CI at least. As far I remember, there is no official board available from MCHP that uses SAMV70. Once SAMV70 it is a subset from SAMV71 (less Ethernet) I`ll suggest to add entries at boards/arm/sam_v71_xult for that purpose. In this case, there will be have 4 targets inside sam_v71_xult board directory.

Hello @nandojve , thank you for the review. I don't fully understand what you would like me to do. Can you clarify what you mean?

Should I copy the board overlay for the sam_v71_xult and make a new, fake one for an imaginary board? In particular, I am not sure about:

1) What to name it
2) Where to put the Kconfig options
3) Do I make a new board name, defconfig, dts, pin-ctrl, and yaml file for a imaginary board?

nandojve commented:
Hi @iperry ,

> Should I copy the board overlay for the sam_v71_xult and make a new, fake one for an imaginary board? In particular, I am not sure about:
> 
> 1. What to name it
> 2. Where to put the Kconfig options
> 3. Do I make a new board name, defconfig, dts, pin-ctrl, and yaml file for a imaginary board?

1. Zephyr identify boards by content inside a board directory, not necessarily the directory name itself. When you run `west boards` you already see two boards for `sam_v71_xult`: _sam_v71_xult_ and _sam_v71b_xult_. Now you need add two other boards:  _sam_v70_xult_ and _sam_v70b_xult_.

2-3. There is no necessity to create a new board directory. You need just create _sam_v70[b]_xult.dts__, _sam_v70[b]_xult.yaml__ and _sam_v7[b]_xult_defconfig__ files to make a new board available to Zephyr. You need edit `Kconfig.foo`, `openocd.cfg` and `index.rst` files. In the case of documentation you should mention the support of SAMV70 and clarify that only network is not available in this specific Zephyr board. You should rename `sam_v71_xult-pinctrl.dtsi` as `sam_v7x_xult-pinctrl-common.dtsi` to be included by two new files `sam_v71_xult-pinctrl.dtsi` and `sam_v70_xult-pinctrl.dtsi`. We need to do this to make sure we are including correct dt-bindings pinctrl file. You should do a similar process with `sam_v71_xult-common.dtsi` and move `gmac` node to a new `sam_v71_xult-common.dtsi` file. I think it may not be necessary create a `sam_v70_xult-common.dtsi` file.

Note: It is possible to use Kconfig variables to instruct OpenOCD, see below as reference: https://github.com/zephyrproject-rtos/zephyr/blob/d84989a718272f29f3a19d4ea99afc7a6aa73211/boards/common/openocd.board.cmake#L7-L9

iperry committed:
boards: sam_v70(b)_xult: add SAM V70 Xplained test target

Introduce two new variations (sam_v70_xult and sam_v70b_xult) of the SAM
V71 Xplained Ultra development board. While these are not actually
manufactured variations, they exist to test the SAM V70 support. These
boards are configured with an ATSAMV70Q20(b) processor and are
functionally identical to the SAM V71 versions, minus Ethernet.

Signed-off-by: Perry Hung <perry@genrad.io>

iperry commented:
Hi @nandojve , thanks for the clarification.

I've prepared two test targets that I believe implements what you asking for. I introduced two boards, ran twister tests, and fixed any Kconfig errors that I could find.

nandojve commented on drivers/can/Kconfig.sam:
```diff
@@ -8,3 +8,15 @@ config CAN_SAM
 	bool "Atmel SAM CAN driver"
 	default $(dt_compat_enabled,$(DT_COMPAT_ATMEL_SAM_CAN))
 	select CAN_MCAN
+
+if CAN_SAM
+
+config CAN_MAX_FILTER
+	int "Maximum number of concurrent active filters"
+	default 5
+	range 1 32
+	help
+	  Defines the array size of the callback/msgq pointers.
+	  Must be at least the size of concurrent reads.
+
+endif # CAN_SAM
```
This seems to be a separated commit or could even be on a different PR.
Could you clarify?

nandojve commented on drivers/flash/Kconfig.sam:
```diff
@@ -10,6 +10,6 @@ config SOC_FLASH_SAM
 	select FLASH_HAS_DRIVER_ENABLED
 	select MPU_ALLOW_FLASH_WRITE if ARM_MPU
 	depends on SOC_SERIES_SAME70 || \
-		   SOC_SERIES_SAMV71
+		   SOC_SERIES_SAMV71 || SOC_SERIES_SAMV70
```
Keep pattern of one per line were V70 came before V71

nandojve commented on soc/arm/atmel_sam/samv70/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,71 @@
+# are used).
+#
+config NUM_IRQS
+	default 74 if SOC_ATMEL_SAMV70_REVB
```
71 instead 74

nandojve commented on soc/arm/atmel_sam/samv70/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,71 @@
+#
+config NUM_IRQS
+	default 74 if SOC_ATMEL_SAMV70_REVB
+	default 71
```
69 instead 71

nandojve commented on boards/arm/sam_v71_xult/Kconfig.defconfig:
```diff
@@ -44,3 +44,10 @@ config ETH_SAM_GMAC
 endif # NETWORKING
 
 endif # BOARD_SAM_V71_XULT
+
+if BOARD_SAM_V70_XULT
+
+config BOARD
+	default "sam_v70_xult"
+
+endif # BOARD_SAM_V70_XULT
```
Let's try keep less specialized always first.

nandojve commented on boards/arm/sam_v71_xult/Kconfig.board:
```diff
@@ -7,3 +7,7 @@
 config BOARD_SAM_V71_XULT
 	bool "Atmel SMART SAM V71 Xplained Ultra Board"
 	depends on SOC_PART_NUMBER_SAMV71Q21 || SOC_PART_NUMBER_SAMV71Q21B
+
+config BOARD_SAM_V70_XULT
+	bool "Atmel SMART SAM V70 Xplained Ultra Board"
+	depends on SOC_PART_NUMBER_SAMV70Q20 || SOC_PART_NUMBER_SAMV70Q20B
```
Let's try keep less specialized always first.

nandojve commented on boards/arm/sam_v71_xult/doc/index.rst:
```diff
@@ -1,7 +1,7 @@
 
-SAM V71(B) Xplained Ultra
-#########################
+SAM V70/V71 (B) Xplained Ultra
+##############################
```
V70/V71(B)

nandojve commented on dts/arm/atmel/samv70.dtsi:
```diff
@@ -0,0 +1,9 @@
+
+#include <atmel/same70.dtsi>
+
+#include "dma_atmel_samv71.h"
```
You should exclude gmac node here.

nandojve commented on dts/arm/atmel/samv70b.dtsi:
```diff
@@ -0,0 +1,9 @@
+
+#include <atmel/same70b.dtsi>
+
+#include "dma_atmel_samv71.h"
```
You should exclude gmac node here.

nandojve reviewed (changes requested):
Hi @iperry,

Almost there. See my comments.

nandojve commented:
Hi @iperry ,

I was wondering if you have time to address our concerns. It will be nice have it for 3.2
